src/routers/alpha-router/gas-models/mixedRoute/on-chain-heuristic-gas-model.ts

```diff
@@ -0,0 +1,448 @@
+        );
+        return {
+          gasEstimate: baseGasUse,
+          gasCostInToken: CurrencyAmount.fromRawAmount(token, 0),
+          gasCostInUSD: CurrencyAmount.fromRawAmount(usdToken, 0),
+        };
+      }
+
+      const token0 = nativePool.token0.address == nativeCurrency.address;
+
+      // returns mid price in terms of the native currency (the ratio of quoteToken/nativeToken)
+      const nativeTokenPrice = token0
+        ? nativePool.token0Price
+        : nativePool.token1Price;
+
+      let gasCostInTermsOfQuoteToken: CurrencyAmount;
+      try {
+        // native token is base currency
+        gasCostInTermsOfQuoteToken = nativeTokenPrice.quote(
+          totalGasCostNativeCurrency
+        ) as CurrencyAmount;
+      } catch (err) {
+        log.info(
+          {
+            nativeTokenPriceBase: nativeTokenPrice.baseCurrency,
+            nativeTokenPriceQuote: nativeTokenPrice.quoteCurrency,
+            gasCostInEth: totalGasCostNativeCurrency.currency,
+          },
+          'Debug eth price token issue'
+        );
+        throw err;
+      }
+
+      // true if token0 is the native currency
+      const token0USDPool = usdPool.token0.address == nativeCurrency.address;
+
+      // gets the mid price of the pool in terms of the native token
+      const nativeTokenPriceUSDPool = token0USDPool
+        ? usdPool.token0Price
+        : usdPool.token1Price;
+
+      let gasCostInTermsOfUSD: CurrencyAmount;
+      try {
+        gasCostInTermsOfUSD = nativeTokenPriceUSDPool.quote(
+          totalGasCostNativeCurrency
+        ) as CurrencyAmount;
+      } catch (err) {
+        log.info(
+          {
+            usdT1: usdPool.token0.symbol,
+            usdT2: usdPool.token1.symbol,
+            gasCostInNativeToken: totalGasCostNativeCurrency.currency.symbol,
+          },
+          'Failed to compute USD gas price'
+        );
+        throw err;
+      }
+
+      return {
+        gasEstimate: baseGasUse,
+        gasCostInToken: gasCostInTermsOfQuoteToken,
+        gasCostInUSD: gasCostInTermsOfUSD!,
+      };
+    };
+
+    return {
+      estimateGasCost: estimateGasCost.bind(this),
+      calculateL1GasFees,
+    };
+  }
+
+  private estimateGas(
+    routeWithValidQuote: MixedRouteWithValidQuote,
+    gasPriceWei: BigNumber,
+    chainId: ChainId
+  ) {
+    const totalInitializedTicksCrossed = BigNumber.from(
+      Math.max(1, _.sum(routeWithValidQuote.initializedTicksCrossedList))
+    );
+    /**
+     * A V3 hop was 1:1 for every pool (i.e. if we went from pool A to pool B it is 1 hop)
+     * Similarly with V2
+     *
+     * Naive approach: every v3 pool is a hop, and counted as 80_000 gas
+     *                 every v2 pool is a hop, and counted as 20_000 gas
+     * Does not matter if v2->v3 or v3->v2
+     * - we will test and see if this hold true
+     */
```

Comment:
tldr: I will do some simulations on tenderly(?) to test these assumptions and to build a correct gas model